karapace/protobuf/io.py

```diff
@@ -0,0 +1,228 @@
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import subprocess
+
+ZERO_BYTE = b'\x00'
+
+logger = logging.getLogger(__name__)
+
+
+def calculate_class_name(name: str) -> str:
+    return "c_" + hashlib.md5(name.encode('utf-8')).hexdigest()
+
+
+def check_props(schema_one, schema_two, prop_list):
+    try:
+        return all(getattr(schema_one, prop) == getattr(schema_two, prop) for prop in prop_list)
+    except AttributeError:
+        return False
+
+
+def match_schemas(writer_schema: ProtobufSchema, reader_schema: ProtobufSchema) -> bool:
+    # TODO (serge): schema comparison by fields required
+
+    return str(writer_schema) == str(reader_schema)
+
+
+def read_varint(bio: BytesIO) -> int:
+    """Read a variable-length integer.
+
+    :returns: Integer
+    """
+    varint = 0
+    read_bytes = 0
+
+    while True:
+        char = bio.read(1)
+        if len(char) == 0:
+            if read_bytes == 0:
+                return 0
+            raise EOFError(f"EOF while reading varint, value is {varint} so far")
+
+        byte = ord(char)
+        varint += (byte & 0x7F) << (7 * read_bytes)
+
+        read_bytes += 1
+
+        if not byte & 0x80:
+            return varint
+
+
+def read_indexes(bio: BytesIO):
+    try:
+        size: int = read_varint(bio)
+    except EOFError:
+        # TODO: change exception
+        raise IllegalArgumentException("problem with reading binary data")
+    if size == 0:
+        return [0]
+    return [read_varint(bio) for _ in range(size)]
+
+
+def find_message_name(schema: ProtobufSchema, indexes: List[int]) -> str:
+    result: list = []
+    types = schema.proto_file_element.types
+    for index in indexes:
+        try:
+            message = types[index]
+        except IndexError:
+            raise IllegalArgumentException(f"Invalid message indexes: {indexes}")
+
+        if message and isinstance(message, MessageElement):
+            result.append(message.name)
+            types = message.nested_types
+        else:
+            raise IllegalArgumentException(f"Invalid message indexes: {indexes}")
+
+    # for java we also need package name. But in case we will use protoc
+    # for compiling to python we can ignore it at all
+    return ".".join(result)
+
+
+def get_protobuf_class_instance(schema: ProtobufSchema, class_name: str) -> Any:
+    directory = config.DEFAULTS["protobuf_runtime_directory"]
```

Comment:
Maybe pass configuration as an argument? As it is it requires the project code to be modified.

Reply:
ok